File: 2022/05/05.kt
// much of this is needlessly functional, but I wanted to try to use this feature as much as possible on this day
fun main() {
    val (tower, moves) = generateSequence(::readlnOrNull).joinToString("\n").split("\n\n")

    val stacks: MutableList<MutableList<Char>> = MutableList(9) { mutableListOf() }

    tower
        .split("\n")
        .reversed()
        .drop(1)
        .forEach { for (i in 1..it.length step 4) if (it[i] != ' ') stacks[i / 4].add(it[i]) }

    val nums = moves
        .split("\n")
        .map { move -> move.filter { !it.isLetter() } }
        .map { it.split("  ") }
        .map { move -> move.map { it.trim().toInt() } }

    for ((n, start, end) in nums) {
        val startStack = stacks[start - 1]
        val endStack = stacks[end - 1]
        val len = startStack.size

        // use this line instead of next for star 1
        // endStack.addAll(startStack.subList(len-n,len).reversed())
        endStack.addAll(startStack.subList(len - n, len))
        stacks[start - 1] = startStack.dropLast(n).toMutableList()
    }

    val arrangement = stacks.map { it.last() }.joinToString("")

    println(arrangement)

    // assert(arrangement == "TQRFCBSJJ") { "First answer is wrong." }
    assert(arrangement == "RMHFJNVFP") { "Second answer is wrong." }
}

File: 2022/05/05.py
import sys

lines = sys.stdin.read().split("\n")

stacks = [[] for i in range(9)]

first = 0

for line in lines:
    first += 1
    if line[1] != '1':
        elems = line[1::4]
        print(elems)
        for i in range(len(elems)):
            if elems[i] != ' ':
                stacks[i].append(elems[i])
    else:
        break

stacks = list(map((lambda l: l[::-1]), stacks))
print(stacks)

for line in lines[first+1:-1]:
    line = line.split()
    print(line)
    amount, s1, s2 = int(line[1]), int(line[3]), int(line[5])

    move = stacks[s1-1][-amount:]#[::-1] #(add for star 1)
    stacks[s1 - 1] = stacks[s1 - 1][:-amount]
    stacks[s2 - 1] = stacks[s2 - 1] + move
    print(amount, s1, s2)
    print(stacks)

for stack in stacks:
    print(stack[-1], end="")
print()

